mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/DeletedMessageZipper.java

```diff
@@ -0,0 +1,73 @@
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.vault;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.stream.Stream;
+
+import org.apache.commons.compress.archivers.zip.ExtraFieldUtils;
+import org.apache.commons.compress.archivers.zip.ZipArchiveEntry;
+import org.apache.commons.compress.archivers.zip.ZipArchiveOutputStream;
+import org.apache.commons.io.IOUtils;
+import org.apache.james.mailbox.backup.MessageIdExtraField;
+import org.apache.james.mailbox.backup.SizeExtraField;
+import org.apache.james.mailbox.model.MessageId;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class DeletedMessageZipper {
+
+    DeletedMessageZipper() {
+        ExtraFieldUtils.register(MessageIdExtraField.class);
+        ExtraFieldUtils.register(SizeExtraField.class);
+    }
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(DeletedMessageZipper.class);
+
+    void zip(Stream<DeletedMessageWithContent> deletedMessages, OutputStream outputStream) throws IOException {
+        ZipArchiveOutputStream zipOutputStream = new ZipArchiveOutputStream(outputStream);
+
+        deletedMessages
+            .forEach(deletedMessage -> putMessageToEntry(zipOutputStream, deletedMessage));
+
+        zipOutputStream.finish();
+    }
+
+    private void putMessageToEntry(ZipArchiveOutputStream zipOutputStream, DeletedMessageWithContent fullMessage) {
+        try {
+            DeletedMessage message = fullMessage.getDeletedMessage();
+            MessageId messageId = message.getMessageId();
+            ZipArchiveEntry archiveEntry = (ZipArchiveEntry) zipOutputStream.createArchiveEntry(new File(messageId.serialize()), messageId.serialize());
+
+            archiveEntry.addExtraField(new MessageIdExtraField(messageId.serialize()));
+            archiveEntry.addExtraField(new SizeExtraField(message.getSize()));
+            zipOutputStream.putArchiveEntry(archiveEntry);
+
+            IOUtils.copy(fullMessage.getContent(), zipOutputStream);
+
+            zipOutputStream.closeArchiveEntry();
+        } catch (IOException e) {
+            LOGGER.warn("Exception happens when putting deleted message {} to the output stream",
```

Comment:
At this point you probably have no idea what the state of the OutputStream is. Why do you think catching the Exception and write a log will allow to go ahead with remaining entries?

Reply:
I just have a look to `ZipArchiveOutputStream` implementations, exception thrown are mostly for checking outputstream ending/closing. Then, what do you think if I will propagate exceptions from  `ZipArchiveOutputStream` methods, and keep continuing the work if non `ZipArchiveOutputStream` stuffs get exceptions? Like IOUtils.copy(stream1, stream2)...

Reply:
In all cases, the state is unknown AFAIU so you should throw